modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryMemoryTracker.java

```diff
@@ -203,74 +224,223 @@ public boolean isOffloadingEnabled() {
         return offloadingEnabled;
     }
 
+    /** {@inheritDoc} */
+    @Override public synchronized void swap(long size) {
+        assert size >= 0;
+
+        if (size == 0)
+            return;
+
+        checkClosed();
+
+        if (parent != null)
+            parent.swap(size);
+
+        writtenOnDisk += size;
+        totalWrittenOnDisk += size;
+        maxWrittenOnDisk = Math.max(maxWrittenOnDisk, writtenOnDisk);
+    }
+
+    /** {@inheritDoc} */
+    @Override public synchronized void unswap(long size) {
+        assert size >= 0;
+
+        if (size == 0)
+            return;
+
+        checkClosed();
+
+        if (parent != null)
+            parent.unswap(size);
+
+        writtenOnDisk -= size;
+    }
+
     /**
-     * @return {@code True} if closed, {@code False} otherwise.
+     * @return {@code true} if closed, {@code false} otherwise.
      */
-    public synchronized boolean closed() {
+    @Override public boolean closed() {
         return closed;
     }
 
     /** {@inheritDoc} */
-    @Override public synchronized void close() {
+    @Override public void close() {
         // It is not expected to be called concurrently with reserve\release.
         // But query can be cancelled concurrently on query finish.
-        if (closed)
+        if (!closing.compareAndSet(false, true))
```

Comment:
Actually, there is no need to have reference to AtomicBoolean object.
It can be replaced with primitive and CAS can be replaced with Unsafe field update operation.